File: src/lib/api/galleryMaterials.ts
import {
  APP_SERVER_METHOD_GALLERY_MATERIAL_GET,
  APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_IMAGE_CATEGORY,
  APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_LAYOUT_CATEGORY,
  APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_MOOD,
  APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_CREATE,
  APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_DELETE,
  APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_UPDATE,
  createAppServerClient,
} from "./appServer";
import type {
  ColorMood,
  CreateGalleryMetadataRequest,
  GalleryMaterial,
  GalleryMaterialMetadata,
  ImageCategory,
  LayoutCategory,
} from "@/types/gallery-material";

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === "object";
}

function assertNotErrorEnvelope(command: string, value: unknown): void {
  const hasErrorEnvelope =
    (isRecord(value) && "error" in value) ||
    (Array.isArray(value) &&
      value.some((item) => isRecord(item) && "error" in item));
  if (hasErrorEnvelope) {
    throw new Error(`${command} returned an error envelope`);
  }
}

function unwrapResponseField<T>(
  command: string,
  response: unknown,
  field: string,
): T {
  assertNotErrorEnvelope(command, response);
  if (!isRecord(response) || !(field in response)) {
    throw new Error(`${command} did not return ${field}`);
  }
  const result = response[field];
  assertNotErrorEnvelope(command, result);
  return result as T;
}

function isOptionalString(value: unknown): value is string | undefined {
  return value === undefined || typeof value === "string";
}

function isOptionalFiniteNumber(value: unknown): value is number | undefined {
  return (
    value === undefined || (typeof value === "number" && Number.isFinite(value))
  );
}

function isStringArray(value: unknown): value is string[] {
  return (
    Array.isArray(value) && value.every((item) => typeof item === "string")
  );
}

function isGalleryMaterialMetadata(
  value: unknown,
): value is GalleryMaterialMetadata {
  return (
    isRecord(value) &&
    typeof value.materialId === "string" &&
    isStringArray(value.colors) &&
    isOptionalFiniteNumber(value.width) &&
    isOptionalFiniteNumber(value.height) &&
    isOptionalString(value.thumbnail) &&
    isOptionalString(value.colorSchemeJson) &&
    isOptionalFiniteNumber(value.elementCount) &&
    isOptionalString(value.preview) &&
    isOptionalString(value.fabricJson) &&
    typeof value.createdAt === "number" &&
    Number.isFinite(value.createdAt) &&
    typeof value.updatedAt === "number" &&
    Number.isFinite(value.updatedAt)
  );
}

function isGalleryMaterial(value: unknown): value is GalleryMaterial {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    typeof value.projectId === "string" &&
    typeof value.name === "string" &&
    typeof value.type === "string" &&
    isOptionalString(value.filePath) &&
    isOptionalFiniteNumber(value.fileSize) &&
    isOptionalString(value.mimeType) &&
    isOptionalString(value.content) &&
    isStringArray(value.tags) &&
    isOptionalString(value.description) &&
    typeof value.createdAt === "number" &&
    Number.isFinite(value.createdAt) &&
    (value.metadata === undefined || isGalleryMaterialMetadata(value.metadata))
  );
}

function assertGalleryMaterialOrNull(
  command: string,
  value: unknown,
): asserts value is GalleryMaterial | null {
  if (value !== null && !isGalleryMaterial(value)) {
    throw new Error(`${command} did not return gallery material`);
  }
}

function assertGalleryMaterialMetadata(
  command: string,
  value: unknown,
): asserts value is GalleryMaterialMetadata {
  if (!isGalleryMaterialMetadata(value)) {
    throw new Error(`${command} did not return gallery material metadata`);
  }
}

function assertGalleryMaterialList(
  command: string,
  value: unknown,
): asserts value is GalleryMaterial[] {
  if (!Array.isArray(value) || !value.every(isGalleryMaterial)) {
    throw new Error(`${command} did not return gallery materials`);
  }
}

function assertVoidResult(command: string, value: unknown): void {
  if (
    value !== undefined &&
    value !== null &&
    (!isRecord(value) || Object.keys(value).length > 0)
  ) {
    throw new Error(`${command} did not return void result`);
  }
}

export async function getGalleryMaterial(
  materialId: string,
): Promise<GalleryMaterial | null> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_GET;
  const response = await createAppServerClient().getGalleryMaterial({
    materialId,
  });
  assertNotErrorEnvelope(command, response.result);
  const result =
    isRecord(response.result) && "material" in response.result
      ? response.result.material
      : null;
  assertNotErrorEnvelope(command, result);
  assertGalleryMaterialOrNull(command, result);
  return result;
}

export async function createGalleryMetadata(
  request: CreateGalleryMetadataRequest,
): Promise<GalleryMaterialMetadata> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_CREATE;
  const response =
    await createAppServerClient().createGalleryMaterialMetadata(request);
  const result = unwrapResponseField<unknown>(
    command,
    response.result,
    "metadata",
  );
  assertGalleryMaterialMetadata(command, result);
  return result;
}

export async function updateGalleryMetadata(
  materialId: string,
  request: CreateGalleryMetadataRequest,
): Promise<GalleryMaterialMetadata> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_UPDATE;
  const response = await createAppServerClient().updateGalleryMaterialMetadata({
    materialId,
    metadata: { ...request, materialId },
  });
  const result = unwrapResponseField<unknown>(
    command,
    response.result,
    "metadata",
  );
  assertGalleryMaterialMetadata(command, result);
  return result;
}

export async function deleteGalleryMetadata(materialId: string): Promise<void> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_METADATA_DELETE;
  const response = await createAppServerClient().deleteGalleryMaterialMetadata({
    materialId,
  });
  assertVoidResult(command, response.result);
}

export async function listGalleryMaterialsByImageCategory(
  projectId: string,
  category?: ImageCategory | null,
): Promise<GalleryMaterial[]> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_IMAGE_CATEGORY;
  const response =
    await createAppServerClient().listGalleryMaterialsByImageCategory({
      projectId,
      category: category ?? null,
    });
  const result = unwrapResponseField<unknown>(
    command,
    response.result,
    "materials",
  );
  assertGalleryMaterialList(command, result);
  return result;
}

export async function listGalleryMaterialsByLayoutCategory(
  projectId: string,
  category?: LayoutCategory | null,
): Promise<GalleryMaterial[]> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_LAYOUT_CATEGORY;
  const response =
    await createAppServerClient().listGalleryMaterialsByLayoutCategory({
      projectId,
      category: category ?? null,
    });
  const result = unwrapResponseField<unknown>(
    command,
    response.result,
    "materials",
  );
  assertGalleryMaterialList(command, result);
  return result;
}

export async function listGalleryMaterialsByMood(
  projectId: string,
  mood?: ColorMood | null,
): Promise<GalleryMaterial[]> {
  const command = APP_SERVER_METHOD_GALLERY_MATERIAL_LIST_BY_MOOD;
  const response = await createAppServerClient().listGalleryMaterialsByMood({
    projectId,
    mood: mood ?? null,
  });
  const result = unwrapResponseField<unknown>(
    command,
    response.result,
    "materials",
  );
  assertGalleryMaterialList(command, result);
  return result;
}

File: lime-rs/crates/app-server/src/local_data_source/gallery_materials.rs
use app_server_protocol::GalleryMaterialDeleteResponse;
use app_server_protocol::GalleryMaterialFilterParams;
use app_server_protocol::GalleryMaterialListResponse;
use app_server_protocol::GalleryMaterialLookupParams;
use app_server_protocol::GalleryMaterialMetadataCreateParams;
use app_server_protocol::GalleryMaterialMetadataResponse;
use app_server_protocol::GalleryMaterialMetadataUpdateParams;
use app_server_protocol::GalleryMaterialResponse;
use lime_core::database::dao::gallery_material_dao::GalleryMaterialDao;
use lime_core::database::DbConnection;
use lime_core::models::project_model as core_project_model;

pub(crate) fn get_gallery_material(
    db: &DbConnection,
    params: GalleryMaterialLookupParams,
) -> Result<GalleryMaterialResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let material = GalleryMaterialDao::get_gallery_material(&conn, &params.material_id)
        .map_err(|error| error.to_string())?
        .map(protocol_gallery_material_from_core);
    Ok(GalleryMaterialResponse { material })
}

pub(crate) fn create_gallery_material_metadata(
    db: &DbConnection,
    params: GalleryMaterialMetadataCreateParams,
) -> Result<GalleryMaterialMetadataResponse, String> {
    let req = core_gallery_material_metadata_request_from_protocol(params);
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let metadata = GalleryMaterialDao::create(&conn, &req)
        .map_err(|error| error.to_string())
        .map(protocol_gallery_material_metadata_from_core)?;
    Ok(GalleryMaterialMetadataResponse {
        metadata: Some(metadata),
    })
}

pub(crate) fn get_gallery_material_metadata(
    db: &DbConnection,
    params: GalleryMaterialLookupParams,
) -> Result<GalleryMaterialMetadataResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let metadata = GalleryMaterialDao::get(&conn, &params.material_id)
        .map_err(|error| error.to_string())?
        .map(protocol_gallery_material_metadata_from_core);
    Ok(GalleryMaterialMetadataResponse { metadata })
}

pub(crate) fn update_gallery_material_metadata(
    db: &DbConnection,
    params: GalleryMaterialMetadataUpdateParams,
) -> Result<GalleryMaterialMetadataResponse, String> {
    let mut req = core_gallery_material_metadata_request_from_protocol(params.metadata);
    req.material_id = params.material_id.clone();
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let metadata = GalleryMaterialDao::update(&conn, &params.material_id, &req)
        .map_err(|error| error.to_string())
        .map(protocol_gallery_material_metadata_from_core)?;
    Ok(GalleryMaterialMetadataResponse {
        metadata: Some(metadata),
    })
}

pub(crate) fn delete_gallery_material_metadata(
    db: &DbConnection,
    params: GalleryMaterialLookupParams,
) -> Result<GalleryMaterialDeleteResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    GalleryMaterialDao::delete(&conn, &params.material_id).map_err(|error| error.to_string())?;
    Ok(GalleryMaterialDeleteResponse {})
}

pub(crate) fn list_gallery_materials_by_image_category(
    db: &DbConnection,
    params: GalleryMaterialFilterParams,
) -> Result<GalleryMaterialListResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let materials = GalleryMaterialDao::list_gallery_materials_by_image_category(
        &conn,
        &params.project_id,
        params.category.as_deref(),
    )
    .map_err(|error| error.to_string())?
    .into_iter()
    .map(protocol_gallery_material_from_core)
    .collect();
    Ok(GalleryMaterialListResponse { materials })
}

pub(crate) fn list_gallery_materials_by_layout_category(
    db: &DbConnection,
    params: GalleryMaterialFilterParams,
) -> Result<GalleryMaterialListResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let materials = GalleryMaterialDao::list_gallery_materials_by_layout_category(
        &conn,
        &params.project_id,
        params.category.as_deref(),
    )
    .map_err(|error| error.to_string())?
    .into_iter()
    .map(protocol_gallery_material_from_core)
    .collect();
    Ok(GalleryMaterialListResponse { materials })
}

pub(crate) fn list_gallery_materials_by_mood(
    db: &DbConnection,
    params: GalleryMaterialFilterParams,
) -> Result<GalleryMaterialListResponse, String> {
    let conn = db
        .lock()
        .map_err(|error| format!("数据库锁定失败: {error}"))?;
    let mood = params.mood.as_deref().or(params.category.as_deref());
    let materials =
        GalleryMaterialDao::list_gallery_materials_by_mood(&conn, &params.project_id, mood)
            .map_err(|error| error.to_string())?
            .into_iter()
            .map(protocol_gallery_material_from_core)
            .collect();
    Ok(GalleryMaterialListResponse { materials })
}

fn protocol_gallery_material_metadata_from_core(
    metadata: core_project_model::GalleryMaterialMetadata,
) -> app_server_protocol::GalleryMaterialMetadata {
    app_server_protocol::GalleryMaterialMetadata {
        material_id: metadata.material_id,
        image_category: metadata.image_category,
        width: metadata.width,
        height: metadata.height,
        thumbnail: metadata.thumbnail,
        colors: metadata.colors,
        icon_style: metadata.icon_style,
        icon_category: metadata.icon_category,
        color_scheme_json: metadata.color_scheme_json,
        mood: metadata.mood,
        layout_category: metadata.layout_category,
        element_count: metadata.element_count,
        preview: metadata.preview,
        fabric_json: metadata.fabric_json,
        created_at: metadata.created_at,
        updated_at: metadata.updated_at,
    }
}

fn protocol_gallery_material_from_core(
    material: core_project_model::GalleryMaterial,
) -> app_server_protocol::GalleryMaterial {
    app_server_protocol::GalleryMaterial {
        id: material.base.id,
        project_id: material.base.project_id,
        name: material.base.name,
        material_type: material.base.material_type,
        file_path: material.base.file_path,
        file_size: material.base.file_size,
        mime_type: material.base.mime_type,
        content: material.base.content,
        tags: material.base.tags,
        description: material.base.description,
        created_at: material.base.created_at,
        metadata: material
            .metadata
            .map(protocol_gallery_material_metadata_from_core),
    }
}

fn core_gallery_material_metadata_request_from_protocol(
    params: GalleryMaterialMetadataCreateParams,
) -> core_project_model::CreateGalleryMaterialMetadataRequest {
    core_project_model::CreateGalleryMaterialMetadataRequest {
        material_id: params.material_id,
        image_category: params.image_category,
        width: params.width,
        height: params.height,
        thumbnail: params.thumbnail,
        colors: params.colors,
        icon_style: params.icon_style,
        icon_category: params.icon_category,
        color_scheme_json: params.color_scheme_json,
        mood: params.mood,
        layout_category: params.layout_category,
        element_count: params.element_count,
        preview: params.preview,
        fabric_json: params.fabric_json,
    }
}
